how-to-ns/src/com/gfredericks/how_to_ns.clj

```diff
@@ -92,19 +93,26 @@
                       [(join-symbols (first require-clause) x)]
                       (vec (cons (join-symbols (first require-clause) (first x))
                                  (rest x)))))
-                  [require-clause])]
+                  [require-clause])
+        no-refer-all? (not (:allow-refer-all? opts))
+        maybe-placeholderize-all #(if (= :all %) '[???] %)
+        sort-refers (fn [arg]
+                      (if (coll? arg)
+                        (vec (sort arg))
+                        arg))]
     (for [clause clauses
           :let [[ns-sym & kw-args] clause
                 map-args (-> (apply hash-map kw-args)
                              (cond-> (not (:allow-rename? opts))
                                (dissoc :rename))
-                             (cond-> (not (:allow-refer-all? opts))
-                               (update-when :refer #(if (= :all %) '[???] %)))
-                             (update-when :refer
-                                          (fn [arg]
-                                            (if (coll? arg)
-                                              (vec (sort arg))
-                                              arg))))]]
+                             (cond-> no-refer-all?
+                               (update-when :refer maybe-placeholderize-all))
+                             (cond-> no-refer-all?
+                               (update-when :refer-macros maybe-placeholderize-all))
+                             (update-when :refer sort-refers)
+                             (update-when :refer-macros sort-refers)
+                             (update-when (ReaderConditional/create '(:clj :refer :cljs :refer-macros) false) sort-refers)
+                             (update-when (ReaderConditional/create '(:cljs :refer-macros :clj :refer) false) sort-refers))]]
```

Comment:
seems like the `maybe-placeholderize-all` step needs to happen for these two as well

I don't mind if you just add two lines for that; maybe a comment describing that the section could be deduplicated with a `reduce` over the four keys

it might also be nice to sort the `:clj` and `:cljs` clauses in the reader conditional (so there's only one proper way to write them instead of two), but that's not necessary

Reply:
> seems like the maybe-placeholderize-all step needs to happen for these two as well

Added 👍

> it might also be nice to sort the :clj and :cljs clauses in the reader conditional (so there's only one proper way to write them instead of two), but that's not necessary

Good suggestion, I played around with this a bit, but there's no easy way to sort the reader conditional (best i could do was recreate after sorting), which would end up with more code than it's replacing :P